File: easy/178_matrix_rotation.py
from __future__ import print_function
import sys
from math import sqrt

for line in open(sys.argv[1]):
    N = int(sqrt((len(line)+1)/2))
    for i in range(N):
        for j in range(N-1, -1, -1):
            print(line[(N*j+i)*2], end=' ')
    print('')

# fast but more memory
#for line in open(sys.argv[1]):
#    line = line.split()
#    N = int(sqrt(len(line)))
#    for col in zip(*reversed([line[i:i+N] for i in range(0, len(line), N)])):
#        print(' '.join(col), end=' ')
#    print("")

File: easy/178_matrix_rotation.c
#include <stdio.h>
#include <string.h>
#include <math.h>

int main(int argc, const char * argv[]) {
    FILE *file = fopen(argv[1], "r");
    char line[1024];
    char *token;
    int N;

    while (fgets(line, 1024, file)) {
        if (line[strlen(line)-1] == '\n')   /* get rid of new line */
            line[strlen(line)-1] = '\0';
        N = sqrt((strlen(line)+1) / 2);
        for (int i=0; i<N; i++)
            for (int j=N-1; j>=0; j--)
                printf("%c ", line[(N*j+i)*2]);
        printf("\n");
    }

    return 0;
}
